tests/coverage.js

```diff
@@ -0,0 +1,241 @@
+"use strict";
+
+const TestDiscovery = require("./helper/test-discovery");
+const TestCase = require("./helper/test-case");
+const path = require("path");
+const PrismLoader = require("./helper/prism-loader");
+const { BFS, BFSPathToPrismTokenPath } = require("./helper/util");
+const { assert } = require("chai");
+const components = require("../components.json");
+const ALL_LANGUAGES = [...Object.keys(components.languages).filter(k => k !== 'meta')];
+
+
+describe('Pattern test coverage', function () {
+	const testSuite = TestDiscovery.loadAllTests(__dirname + "/languages");
+
+	/**
+	 * @type {Map<string, PatternData>}
+	 * @typedef PatternData
+	 * @property {RegExp} pattern
+	 * @property {string} language
+	 * @property {Set<string>} from
+	 * @property {RegExpExecArray[]} matches
+	 */
+	const patterns = new Map();
+
+	/**
+	 * @param {string | string[]} languages
+	 * @returns {import("./helper/prism-loader").PrismInstance}
+	 */
+	function createInstance(languages) {
+		const Prism = PrismLoader.createInstance(languages);
+
+		BFS(Prism.languages, (path, object) => {
+			const { key, value } = path[path.length - 1];
+			const tokenPath = BFSPathToPrismTokenPath(path);
+
+			if (Object.prototype.toString.call(value) == '[object RegExp]') {
+				const regex = makeGlobal(value);
+				object[key] = regex;
+
+				const patternKey = String(regex);
+				let data = patterns.get(patternKey);
+				if (!data) {
+					data = {
+						pattern: regex,
+						language: path[1].key,
+						from: new Set([tokenPath]),
+						matches: []
+					};
+					patterns.set(patternKey, data);
+				} else {
+					data.from.add(tokenPath);
+				}
+
+				regex.exec = string => {
+					let match = RegExp.prototype.exec.call(regex, string);
+					if (match) {
+						data.matches.push(match);
+					}
+					return match;
+				};
+			}
+		});
+
+		return Prism;
+	}
+
+	describe('Register all patterns', function () {
+		it('all', function () {
+			this.slow(10 * 1000);
+			// This will cause ALL regexes of Prism to be registered in the patterns map.
+			// (Languages that don't have any tests can't be caught otherwise.)
+			createInstance(ALL_LANGUAGES);
+		});
+	});
+
+	describe('Run all language tests', function () {
```

Comment:
Are we re-running all of the language tests? We can't wrap the current test run?

Reply:
> We can't wrap the current test run?

This will only work if *all* language tests are run. 

But in the end, coverage was easier to implement by rerunning all tests. The only point against this is performance, right?

Reply:
Performance & code duplication.

Can we add the coverage checking to the base test suite?

Reply:
I honestly don't think that's a good idea.

Coverage only uses our test files to gather data. There is no guarantee that this won't change/be expanded in the future. I don't want to marry two simple systems only to get one complex system.

If code dup is the problem, then I'll add a helper function that runs the whole test suite for us. The current test runner will then only parse CLI args and call the helper function. This might actually be the best solution since coverage doesn't actually care whether the test cases pass.

Reply:
lol I don't know why it bothers me so much that we're running the test suite in one file, then rerunning the whole suite again in another file to collect coverage. Running the tests & collecting coverage have always gone hand-in-hand in other cases (e.g. running `nyc mocha` or `jest --coverage`), so my inclination is to combine or connect them, less from a perf or code dupe perspective and more of an testing arch perspective.

That said, strong opinion weakly held. I'm not sure I fully understand the downside here.

Reply:
Yeah, I get your perspective. We usually use awesome test coverage tools like [Istanbul](https://istanbul.js.org/). These tools hook directly into NodeJS, so they can get detailed information about what functions/statements/branches are called/executed/takes and how many times. But getting this usage information it's not so easy for Prism.

Our goal is to figure out what patterns contributed to the token streams produced by the test files. Basically, we want to know what tokens each regex produced.

Right now, the coverage test *approximates* this by running all test files and remembering all matches of all (unique) regexes. This is not a correct solution as it overestimates the actual usage of our regexes. I.e. regex matches might lead to tokens that are subsequently discarded because of other greedy patterns. The correct solution is to examine the produced token stream and discard all regex matches that didn't produce a token present in the final output token stream of our test files.
The only reason I haven't implemented the correct solution is that it is hard to do and the approximation seemed good enough for now.

This is why I view the two tests as distinct. They have different goals but achieve them through similar methods, at least for now.